modules/docs/markdown/06-guides/testing.md

```diff
@@ -0,0 +1,41 @@
+
+## How we Test Smithy4s
+
+It is important to understand how Smithy4s works so you can begin to conceptualize how we test Smithy4s and its interpreters. We will not go into much detail on this point, but the main ways we test these are through:
+
+- Unit tests on the interpreters. These tests isolate the interpreters directly and test their functionality. An example of this is the Jsoniter interpreter. We are able to isolate its functionality and test that it is working as intended.
+- Protocol tests. These tests are auto-generated from Smithy-defined traits. They are sort of like integration tests in the sense that they test all the relevant layers of a smithy4s application by connecting, for example, the HTTP layer with the JSON serialization layer. The main goal of these tests is to make sure that Smithy4s is handling edge cases properly and overall conforms to the protocol(s) we have defined. For more on protocol testing, you can check out the [Smithy documentation](https://smithy.io/2.0/additional-specs/http-protocol-compliance-tests.html).
+
+## Testing your API
+
+Now the question that remains is what all of this means for your applications which use Smithy4s. Our basic recommendation is going to come down to: test what you feel needs to be tested, but make sure you understand _what_ it is that you are testing.
+
+For example, we would not generally recommend that you test the Smithy4s interpreters in your code (for example the http4s routes that you get when you use `SimpleRestJsonBuilder`). One of the main appeals of Smithy4s is that it abstracts away concerns such as request routing and the encoding/decoding of payloads. This is why we strive to be thorough in our testing: so you don't need to worry about testing whether or not Smithy4s is doing its job. That being said, if you don't trust Smithy4s and want to test it anyway, that is totally fine too. Ultimately Smithy4s as a project will benefit if you do test it and find any bugs that we can then fix. So in any case, we are supportive of your decision.
+
+Another reason you may think of testing the Smithy4s interpreters is to make sure that their implementation, such as the way they encode something in JSON, lines up with some requirements that you have. This is fair enough, but realize that the thing you are really testing here is your understanding of how Smithy4s works. That may be worth your while, or it may not be. The decision is ultimately up to you. If this is your reason for testing, you may want to explore contract-based testing as outlined below.
```

Comment:
**thought:** I think there's another reason why people do this: 

They don't trust they wrote the right Smithy specs (or that they didn't make a mistake in refactoring them).

Let me explain:

imagine I decided with my colleagues that we're going to return a payload with a top-level string.

I write the spec:

```smithy
@http(...)
operation Foo {
  output := {
    sth: String
  }
}
```

I add a smithy4s implementation of this operation and don't write route-level tests for it. Turns out I messed up, the response actually looks like this:

```json
{
  "sth": "foo"
}
```

instead of what I expected to be `"foo"`.

---

so this is why the users would write a test on the routes, hitting the right URL and checking if the response is 

```scala
val response = Json.fromString("foo")
```

. Now, I don't think these are good tests because they run way too late after introducing the mistake, and a better solution (on a similar level of setup complexity) would be writing compliance tests. Then, if they want to be hardcore about it, contract testing...

Reply:
I'd say this sounds like what you described as schema testing, but I've seen it also involve testing error responses based on the request.

Reply:
Yeah I guess this falls into the category that I talked about of "testing your own understanding of Smithy/Smithy4s" which I think there is value in.. It just _feels_ like it shouldn't be necessary. Though I do agree that there is a certain sort of regression that it could prevent.

I think if we were to recommend that people do this kind of testing, we may want to provide something that makes it easier out of the box. Like for example we could have a validator that checks that the `examples` trait (or a new trait) values match what a request/response would actually look like for the operation given the protocol at hand.  Kind of like compliance tests, but here we are really checking that the encoding is what the user expects, not that smithy4s will produce the expected encoding. If any of that makes sense 😆